### ISSUE TITLE
Consistent node type categorization

### ISSUE THREAD

**Issue by michael** [substance/substance#167]
We have the following node type categories, which are sometimes defined via inheritance chain, sometimes via a static property definition. Let's see if we can unify this to be able to better reason about in which category a given node is. However let's find use-cases so we don't engineer blindly what could be useful..

So far we are aware of the following categories:

**Text**

Any text-ish node, such as paragraph, heading, caption etc.

**Block**

Parent is a container, so it's kind of a root node. E.g. image, paragraph, heading.

**Container**

Manages a list of Block nodes (List, Paragraph, Codeblock, Blockquote)

**Annotation**

Overlays characters of a text property. E.g. Strong, Link, etc.

**Inline**

An property annotation that also has content. E.g. a dynamically labelled FigureCaption where the label is the content. But this could also be an InlineImage. Or InlineFormula. Inline nodes wrap a fake character, that just says there is some content here that you can not edit.

**ContainerAnnotation**

Special annotation that sticks on a container rather then a text property.

We can not describe this entirely via inheritance chain. Because for instance a Paragraph would inherit from Block AND Text. 

``` js
Node.static = {
  isText: false,
  isBlock: false,
  isInline: false,
  isAnnotation: false,
  isContainer: false,
  isContainerAnnotation: false
};
```

We can have a class hierarchy like this and we just set the static properties accordingly. Except for isBlock, which needs to be set manually in all block implementations, much like it is the case already using blockType = true)

```
- Node
   - Text
   - Annotation
      - Inline
   - ContainerAnnotation
   - Container
```

**Comment by obuchtala** [comment]
'Container' is a bit unclear. List for example is not a container, it is a nested node type, same as Table.
And don't want to create confusion with the exisiting Container node.

**Comment by michael** [comment]
List would be just a block in that terminology. The question is what is a ListItem or TableRow? It seems we can not say a node is either a block or inline.. there's more...

**Comment by michael** [comment]
In need this checking stuff in different places. Considering that the node refactor is now in.. can we take care of this now?

In SwitchTextType I need to check if the node is a Block and a Text element. I have the node instance in this case. The options I have currently:

``` js
node instanceOf TextNode // need to require TextNode to do that
node.constructor.static.blockType // a bit combersome
node instanceof TextBlock // combined check but still needs require
```

Suggestions for more straight-forward checking:

``` js
node.isBlock();
node.isText();
node.isInline(); // currently external
node.isAnnotation();

// this we would need to define on the DocumentNode to have it as an interface, and each 
// function would return false by default. Howevver you mentioned that you don't like to put 
// it there since DocumentNode could be kept more clean and generic, not knowing about blocks and text things. 
// So alternatively we could provide an `is()` method that is abstract and can be implemented on the classes using the info form static.

node.is('block')
node.is('text')
node.is('inline')
node.is('annotation')

// On the node class maybe more like this:
NodeClass.static.is('block')
```

I think what we need to do is announcing these properties in the API docs and then users don't need to think so much on how to check for a certain category.

**Comment by obuchtala** [comment]
I think isBlock(), isInline(), and isAnnotation() should be on the DocumentNode interface as these are essential core concepts.

With text, you can already do `node.isInstanceOf('text')` which works on the schema.

**Comment by michael** [comment]
what would speak again isText() and isTextBlock() too? It's a bit redundant i know.. but somehow it's odd to sometimes have a prop and sometimes check via inheritance chain.

**Comment by obuchtala** [comment]
isText() ok... but that is the last one ;)

IsTextBlock() === isText() && isBlock()

**Comment by michael** [comment]
hehe ahja! ;) deal! I will do that then now. Thanks!

**Comment by michael** [comment]
oh and for static... let's just call it isBlock istead of blockType so it matches with the function name?

**Comment by obuchtala** [comment]
ok.

**Comment by obuchtala** [comment]
Hmmm.. maybe 'isBlock' is not a good idea though, as it would be confusing to do `paragraph.isBlock()` but `Paragraph.static.isBlock`?

**Comment by obuchtala** [comment]
dunno.

**Comment by obuchtala** [comment]
Just wanted to say this.

**Comment by obuchtala** [comment]
Maybe we can also achieve to rename 'external' to 'isInline'?

**Comment by michael** [comment]
is it confusing? `paragraph.isBlock()` and `Paragraph.static.isBlock` ? it's not so bad i would say... unless we want a function on static too?

external -> inline yes!

**Comment by michael** [comment]
and strong,emphasis,links etc. are just annotations or could that also have a name in the terminology?

**Comment by obuchtala** [comment]
This I want to call annotation as opposed to inlineNode. I.e., annotation doesn't own the plain text.

**Comment by michael** [comment]
inlineNode is also an annotation, technically. But it's spcial in that it owns the content. can we say it like this?

**Comment by michael** [comment]
we also have some confusion in naming because for instance in the converter we call all annotation converters 'inlineConverters'.

**Comment by obuchtala** [comment]
Yes, only implementation wise. But not semantically.

**Comment by obuchtala** [comment]
Hmmm.. agreed.

**Comment by obuchtala** [comment]
Maybe I have to reconsider my categorization?

What is your suggestion?

**Comment by michael** [comment]
I mean we just must not abuse the word 'inline' for other things than this node type thingee.. where it is at home the most in my opinion.

maybe call the categorization just 'annotationConverters'?

**Comment by obuchtala** [comment]
Yep, agreed. Annotation is implementation wise the more generic.

**Comment by michael** [comment]
Should we have an `Inline` class that inherits from PropertyAnnotation? currently propertyAnnotation sets isInline true.. i would have that false by default?

**Comment by obuchtala** [comment]
Agreed.

**Comment by michael** [comment]
Inline.js or InlineNode.js?

**Comment by obuchtala** [comment]
InlineNode

**Comment by michael** [comment]
container annos handlst ja no gar ned im converter oder?

i würd dann den move auf _annotationConverters machen.. frag mi nur wo die containerannos da dann reinpassen.

**Comment by obuchtala** [comment]
Stimmt!

**Comment by michael** [comment]
also wird das eher a ganz eigener bin _containerAnnotations im converter... ? Weil dann würd i glei sagen _propertyAnnotations. inlineNode is ja technically auch a propertyAnnotation.

**Comment by obuchtala** [comment]
yup, ich denk

**Comment by michael** [comment]
Seit wann redma hier deutsch? ;) sorry.

I mach `_blockTypeConverters` -> `_blockConverters` und `_inlineTypeConverters` zu `_propertyAnnotationConverters`

**Comment by obuchtala** [comment]
ok.

**Comment by michael** [comment]
Fixed via #264
